curve25519-dalek/src/lemmas/scalar_lemmas_extra.rs

```diff
@@ -0,0 +1,789 @@
+    requires
+        mask == (1u64 << 52) - 1u64,
+        lo[0] == words[0] & mask,
+        lo[1] == ((words[0] >> 52) | (words[1] << 12)) & mask,
+        lo[2] == ((words[1] >> 40) | (words[2] << 24)) & mask,
+        lo[3] == ((words[2] >> 28) | (words[3] << 36)) & mask,
+        lo[4] == ((words[3] >> 16) | (words[4] << 48)) & mask,
+    ensures
+        five_limbs_to_nat_aux(*lo) == (words[0] as nat) + pow2(64) * (words[1] as nat) + pow2(128)
+            * (words[2] as nat) + pow2(192) * (words[3] as nat) + pow2(256) * ((words[4]
+            & 0xf) as nat),
+{
+    let masked_words_sum = ((words[0] & mask) as nat) + pow2(52) * ((((words[0] >> 52) | (words[1]
+        << 12)) & mask) as nat) + pow2(104) * ((((words[1] >> 40) | (words[2] << 24))
+        & mask) as nat) + pow2(156) * ((((words[2] >> 28) | (words[3] << 36)) & mask) as nat)
+        + pow2(208) * ((((words[3] >> 16) | (words[4] << 48)) & mask) as nat);
+
+    let unmasked_words_sum = (words[0] as nat) + pow2(64) * (words[1] as nat) + pow2(128) * (
+    words[2] as nat) + pow2(192) * (words[3] as nat) + pow2(256) * ((words[4] & 0xf) as nat);
+
+    let limb1 = (((words[0] >> 52) | (words[1] << 12)) & mask) as nat;
+    let limb2 = (((words[1] >> 40) | (words[2] << 24)) & mask) as nat;
+    let limb3 = (((words[2] >> 28) | (words[3] << 36)) & mask) as nat;
+    let limb4 = (((words[3] >> 16) | (words[4] << 48)) & mask) as nat;
+
+    let w0_low = ((words[0] & mask) as nat);
+    let w0_high = (words[0] >> 52) as nat;
+
+    let w1_low = ((words[1] & (u64::MAX >> 24)) as nat);
+    let w1_high = (words[1] >> 40) as nat;
+    let w2_low = ((words[2] & (u64::MAX >> 36)) as nat);
+    let w2_high = (words[2] >> 28) as nat;
+    let w3_low = ((words[3] & (u64::MAX >> 48)) as nat);
+    let w3_high = (words[3] >> 16) as nat;
+    let w4_low = (words[4] & 0xf) as nat;
+
+    // Limb 1 consists of word 0's top 12 bits and word 1's low 40 bits.
+    assert(limb1 == w0_high + pow2(12) * w1_low) by {
+        let w0 = words[0];
+        let w1 = words[1];
+        let high12 = w0 >> 52;
+        let low40_mask: u64 = u64::MAX >> 24;
+        let low40 = w1 & low40_mask;
+
+        assert((1u64 << 52) - 1u64 == u64::MAX >> 12) by (bit_vector);
+        assert(((w0 >> 52) | (w1 << 12)) & (u64::MAX >> 12) == (w0 >> 52) | ((w1 & (u64::MAX >> 24))
+            << 12)) by (bit_vector);
+        assert((w0 >> 52) < (1u64 << 12)) by (bit_vector);
+        assert((w1 & (u64::MAX >> 24)) <= u64::MAX >> 12) by (bit_vector);
+
+        assert(((w0 >> 52) | ((w1 & (u64::MAX >> 24)) << 12)) < (1u64 << 52)) by (bit_vector);
+        lemma_bit_or_is_plus(high12, low40, 12);
+        assert((low40 << 12) as nat == pow2(12) * (low40 as nat)) by {
+            let shift_u64: u64 = 12;
+            vstd::bits::lemma_u64_mul_pow2_le_max_iff_max_shr(low40, shift_u64, u64::MAX);
+            vstd::bits::lemma_u64_shl_is_mul(low40, shift_u64);
+        }
+    };
```

Comment:
There's a lot of duplication here, for each limb the logic is the same with different constatns, this should be a lemma.